### PR TITLE
[TGUI] AI Integrity Restoration Computer

### PR DESCRIPTION
**Pull request by Shadow-Quill**
## What Does This PR Do

The RD's integrity restoration terminal: Now in TGUI

## Why It's Good For The Game

TGUI. The integrity bar changes color depending on the health of the AI within.
Zeroth laws are shown, and hacked/ion laws show as they would on a AI card (ie: a mishmash of random symbols for the index, that cycles every second or so)

## Images of changes

image

## Changelog
:cl:
add: Converted AI Integrity restoration console to TGUI.
/:cl:

### CODE REVIEW

File: tgui/packages/tgui/interfaces/ai_fixer.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            <h3>No artificial intelligence detected.</h3>
+          </Box>
+        </Window.Content>
+      </Window>
+    );
+  } else {
+    let workingAI = null;
+    if (data.stat == 2 || data.stat == null) {
+      workingAI = false;
+    } else { workingAI = true; }
```
Comment (Kyep) [inline tgui/packages/tgui/interfaces/ai_fixer.js:22]:
```suggestion
    } else { 
      workingAI = true;
    }
```

File: tgui/packages/tgui/interfaces/ai_fixer.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    let integrityColor = null;
+    if (data.integrity >= 75) {integrityColor = 'green';}
+    else if (data.integrity >= 25) {integrityColor = 'yellow';}
+    else {integrityColor = 'red';}
+
```
Comment (Kyep) [inline tgui/packages/tgui/interfaces/ai_fixer.js:28]: Maybe add code here that defines let aiLawBlock; and then loops through lawsList to add stuff to it.
Then maybe use {aiLawBlock} below to include it. That way you're not trying to calculate it in the return() section.

File: tgui/packages/tgui/interfaces/ai_fixer.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          </Box>
+          {/* <LabeledList>
+            for (let i = 0; i < lawsList.length; i++) {
+              <LabeledList.Item label={{data.laws}[1]} />
+            }
+          </LabeledList> */}
+
+          {/* {laws.map(law => (
+            {law.number} {law.law}
+          ))} */}
```
Comment (Kyep) [inline tgui/packages/tgui/interfaces/ai_fixer.js:59]:
```suggestion
```

File: code/game/machinery/computer/aifixer.dm [hunk 4]
```diff
@@ -48,43 +47,43 @@
...
+					occupant.update_revive()
+					occupant.lying = 0
+					update_icon()
+				sleep(10)
+			active = 0
+			add_fingerprint(usr)
+
+		if("wireless")
+			var/wireless = text2num("wireless")
+			if(wireless == 0 || wireless == 1)
```
Comment (Kyep) [inline code/game/machinery/computer/aifixer.dm:78]: What are the possible values for this other than 0 or 1?

Comment (Shadow-Quill) [reply code/game/machinery/computer/aifixer.dm:78]: I guess it was for href checks or something??? No clue, gonna just replace that with a
```
if(x)
    thing.to.change = thing```

**Comment by AffectedArc07** [comment]
Note that the modular consoles still have the old nanoUI AI restorer, `ai_restorer.tmpl`. I dont think this will cause issues but its wroth pointing out

File: tgui/packages/tgui/interfaces/ai_fixer.js (hunk 3 above)
Comment (Shadow-Quill) [reply tgui/packages/tgui/interfaces/ai_fixer.js:59]: Just failed attempts from earlier.

**Comment by AffectedArc07** [comment]
Also it may be worth using `<Section>` in here to add the darker backdrop onto things as seen here
image
Though thats up to you

File: code/game/machinery/computer/aifixer.dm [hunk 5]
```diff
@@ -22,23 +22,22 @@
...
-	ui_interact(user)
+	tgui_interact(user)
 
 /obj/machinery/computer/aifixer/attack_hand(var/mob/user as mob)
-	ui_interact(user)
+	tgui_interact(user)
 
-/obj/machinery/computer/aifixer/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
-	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, force_open)
+/obj/machinery/computer/aifixer/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
```
Comment (AffectedArc07) [inline code/game/machinery/computer/aifixer.dm:30]: `force_open` used to be `1`, so set it to `TRUE` here, not `FALSE`

File: code/game/machinery/computer/aifixer.dm [hunk 6]
```diff
@@ -48,43 +47,41 @@
...
+				occupant.adjustToxLoss(-1, FALSE)
+				occupant.adjustBruteLoss(-1, FALSE)
+				occupant.updatehealth()
+				if(occupant.health >= 0 && occupant.stat == DEAD)
+					occupant.update_revive()
+					occupant.lying = 0
+					update_icon()
+				sleep(10)
+			active = 0
+			add_fingerprint(usr)
```
Comment (AffectedArc07) [inline code/game/machinery/computer/aifixer.dm:76]: Make this an async function. We dont want any `sleep()` calls inside `tgui_act`

File: code/game/machinery/computer/aifixer.dm [hunk 7]
```diff
@@ -48,43 +47,41 @@
...
+
+		if("wireless")
+			occupant.control_disabled = !occupant.control_disabled
+
+		if("radio")
+			occupant.aiRadio.disabledAi = !occupant.aiRadio.disabledAi
+
+	SStgui.update_uis(src)
```
Comment (AffectedArc07) [inline code/game/machinery/computer/aifixer.dm:84]: Replace this with `return TRUE`

File: code/game/machinery/computer/aifixer.dm [hunk 8]
```diff
@@ -48,43 +47,41 @@
...
+				occupant.adjustFireLoss(-1, FALSE)
+				occupant.adjustToxLoss(-1, FALSE)
+				occupant.adjustBruteLoss(-1, FALSE)
+				occupant.updatehealth()
+				if(occupant.health >= 0 && occupant.stat == DEAD)
+					occupant.update_revive()
+					occupant.lying = 0
+					update_icon()
+				sleep(10)
+			active = 0
```
Comment (AffectedArc07) [inline code/game/machinery/computer/aifixer.dm:75]: `FALSE` here

File: code/game/machinery/computer/aifixer.dm [hunk 9]
```diff
@@ -48,43 +47,41 @@
...
+			active = 1
+			while(occupant.health < 100)
+				occupant.adjustOxyLoss(-1, FALSE)
+				occupant.adjustFireLoss(-1, FALSE)
+				occupant.adjustToxLoss(-1, FALSE)
+				occupant.adjustBruteLoss(-1, FALSE)
+				occupant.updatehealth()
+				if(occupant.health >= 0 && occupant.stat == DEAD)
+					occupant.update_revive()
+					occupant.lying = 0
```
Comment (AffectedArc07) [inline code/game/machinery/computer/aifixer.dm:72]: `FALSE` here assuming `lying` is a bool

File: code/game/machinery/computer/aifixer.dm [hunk 10]
```diff
@@ -48,43 +47,41 @@
...
-	if(href_list["radio"])
-		var/radio = text2num(href_list["radio"])
-		if(radio == 0 || radio == 1)
-			occupant.aiRadio.disabledAi = radio
-
-	SSnanoui.update_uis(src)
+		return
+	switch(action)
+		if("fix")
+			active = 1
```
Comment (AffectedArc07) [inline code/game/machinery/computer/aifixer.dm:63]: `TRUE`

File: tgui/packages/tgui/interfaces/AIFixer.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+export const AIFixer = (props, context) => {
+  const { act, data } = useBackend(context);
+  if (data.occupant === null) {
+    return (
+      <Window>
+        <Window.Content>
+          <Box>
+            <h3>No artificial intelligence detected.</h3>
+          </Box>
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/AIFixer.js:13]:
```suggestion
          <Section title="Stored AI">
            <Box bold>
              <h3>No artifical intelligence detected</h3>
            </Box>
          </Section>
```

File: code/game/machinery/computer/aifixer.dm [hunk 12]
```diff
@@ -48,45 +47,46 @@
...
+		occupant.adjustFireLoss(-1, FALSE)
+		occupant.adjustToxLoss(-1, FALSE)
+		occupant.adjustBruteLoss(-1, FALSE)
+		occupant.updatehealth()
+		if(occupant.health >= 0 && occupant.stat == DEAD)
+			occupant.update_revive()
+			occupant.lying = FALSE
+			update_icon()
+		sleep(10)
+	active = FALSE
```
Comment (AffectedArc07) [inline code/game/machinery/computer/aifixer.dm:89]: Add a check here to make sure you cant spam this to rapidly fix the AI, or consume limitless CPU cycles

Comment (Shadow-Quill) [reply code/game/machinery/computer/aifixer.dm:89]: handled by disabling the fix button when fixing
`disabled={integrityFull || data.active}`

File: code/game/machinery/computer/aifixer.dm [hunk 13]
```diff
@@ -48,45 +47,48 @@
...
-		var/radio = text2num(href_list["radio"])
-		if(radio == 0 || radio == 1)
-			occupant.aiRadio.disabledAi = radio
-
-	SSnanoui.update_uis(src)
+		return
+	switch(action)
+		if("fix")
+			if(active) // Prevent from starting a fix while fixing.
+				to_chat(usr, "<span class='warning'>You are already fixing this AI!</span>")
```
Comment (AffectedArc07) [inline code/game/machinery/computer/aifixer.dm:65]: This needs a return or its still going to execute

**Comment by AffectedArc07** [comment]
@Shadow-Quill please see requested changes

**Review by AffectedArc07 — APPROVED** [review]

**Pull request merged**